libnopegl/src/node_colorstats.c

```diff
@@ -0,0 +1,414 @@
+        .blocks         = block,
+        .nb_blocks      = 1,
+        .workgroup_size = {s->group_size, 1, 1},
+    };
+
+    return setup_compute(s, s->crafter_waveform, s->pipeline_compat_waveform, &crafter_params);
+}
+
+/* Phase 3: summary and scale for global histograms */
+static int setup_sumscale_compute(struct colorstats_priv *s, const struct pgcraft_block *block)
+{
+    const struct pgcraft_params crafter_params = {
+        .comp_base      = colorstats_sumscale_comp,
+        .blocks         = block,
+        .nb_blocks      = 1,
+        .workgroup_size = {s->group_size, 1, 1},
+    };
+
+    return setup_compute(s, s->crafter_sumscale, s->pipeline_compat_sumscale, &crafter_params);
+}
+
+static int init_computes(struct ngl_node *node)
+{
+    struct ngl_ctx *ctx = node->ctx;
+    struct colorstats_priv *s = node->priv_data;
+    const struct colorstats_opts *o = node->opts;
+    struct gpu_ctx *gpu_ctx = ctx->gpu_ctx;
+
+    /*
+     * Define workgroup size using limits. We pick a value multiple of the
+     * depth on purpose.
+     *
+     * GL_MAX_COMPUTE_WORK_GROUP_SIZE must be at least [128,128,64] (OpenGLES
+     * 3.1), but GL_MAX_COMPUTE_WORK_GROUP_INVOCATIONS (x*y*z) is only 128.
+     * Meaning that if we pick a work group size X=128, we will have to use Y=1
+     * and Z=1.
+     */
+    const struct gpu_limits *limits = &gpu_ctx->limits;
+    const int max_group_size_x = limits->max_compute_work_group_size[0];
+    ngli_assert(max_group_size_x >= 128);
+    s->group_size = max_group_size_x >= 256 ? 256 : 128;
+    LOG(DEBUG, "using a workgroup size of %d", s->group_size);
+
+    s->pipeline_compat_init     = ngli_pipeline_compat_create(gpu_ctx);
+    s->pipeline_compat_waveform = ngli_pipeline_compat_create(gpu_ctx);
+    s->pipeline_compat_sumscale = ngli_pipeline_compat_create(gpu_ctx);
+    if (!s->pipeline_compat_init || !s->pipeline_compat_waveform || !s->pipeline_compat_sumscale)
+        return NGL_ERROR_MEMORY;
+
+    s->crafter_init     = ngli_pgcraft_create(ctx);
+    s->crafter_waveform = ngli_pgcraft_create(ctx);
+    s->crafter_sumscale = ngli_pgcraft_create(ctx);
+    if (!s->crafter_init || !s->crafter_waveform || !s->crafter_sumscale)
+        return NGL_ERROR_MEMORY;
+
+    const struct pgcraft_block block = {
+        .name     = "stats",
+        .type     = NGLI_TYPE_STORAGE_BUFFER,
+        .stage    = NGLI_PROGRAM_SHADER_COMP,
+        .writable = 1,
+        .block    = &s->blk.block,
+        .buffer   = s->blk.buffer,
+    };
+
+    int ret;
+    if ((ret = setup_init_compute(s, &block)) < 0 ||
+        (ret = setup_waveform_compute(s, &block, o->texture_node)) < 0 ||
+        (ret = setup_sumscale_compute(s, &block)) < 0)
+        return ret;
+
+    return 0;
+}
+
+static int init_block(struct colorstats_priv *s, struct gpu_ctx *gpu_ctx)
+{
+    s->blk.data = NULL; /* We do not have any CPU data */
+
+    /*
+     * std430 layout is needed because the buffer has to be writable, and
```

Comment:
std430 has nothing to do with the buffer being writable nor the variadic declaration. You can do that with std140.

Reply:
Removed, though for some reason it seems that using std140 here is leading to errors such as `Assert buffer->usage & NGLI_BUFFER_USAGE_UNIFORM_BUFFER_BIT @ src/backends/gl/pipeline_gl.c:714`. For some reason it appears the pipeline wants to bind it as an UBO when the layout is std140. I couldn't remember nor find where we have this heuristic.

Reply:
I'll make a patch to fix the heuristic.

Reply:
Ok, so there is no patch needed as the colorstats node do something a bit hacky regarding the buffer allocatation / initialisation (lazy allocated). In order for the std140 to work here, we also need to set NGLI_BUFFER_USAGE_UNIFORM_BUFFER_BIT flags to the block and buffer usage.